### ISSUE TITLE
Rewriting + eta-equality for unit type = incomplete conversion

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#5929]
This is a continuation of #5923. The internal error is fixed, however now definitional equality is no longer transitive:
```agda
{-# OPTIONS --rewriting #-}

open import Agda.Builtin.Bool
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Agda.Builtin.Unit

not : Bool → Bool
not true = false
not false = true

sym : {A : Set} {x y : A} → x ≡ y → y ≡ x
sym refl = refl

trans : {A : Set} {x y z : A} → x ≡ y → y ≡ z → x ≡ z
trans refl refl = refl

transport : {A : Set} (P : A → Set)
          → {x y : A} → x ≡ y → P x → P y
transport P refl p = p

postulate
  f : {X A : Set} → X → A
  g : {X A : Set} → A → X

eq₁ : {A : Set} (a : A) → f {⊤} {A} tt ≡ f (g a)
eq₁ a = refl

postulate
  rew-fg : {X A : Set} (a : A) → f (g {X} a) ≡ a
{-# REWRITE rew-fg #-}

eq₂ : {A : Set} (a : A) → f (g {⊤} a) ≡ a
eq₂ a = refl

eq₃ : {A : Set} (a : A) → f tt ≡ a
eq₃ a = trans (eq₁ a) (eq₂ a)

eq : {A : Set} (x y : A) → x ≡ y
eq x y = trans (sym (eq₃ x)) (eq₃ y)
```
Here the term `eq true false : true ≡ false` computes to `refl`, but obviously `refl` does not have type `true ≡ false` (EDIT: or rather, in this system it does, since every type is definitionally equal to every other type, hence `true ≡ false` is equal to the type `true ≡ true` of `refl`. However, the conversion checker of Agda does not agree with this fact.)

**Comment by mikeshulman** [comment]
Surely in the presence of arbitrary rewrites (especially non-confluent ones), definitional equality was already no longer transitive (at least, not in a way verifiable by the conversion checker)?

**Comment by jespercockx** [comment]
Yeah, the problem is that the goal is to guarantee subject reduction and completeness of conversion checking assuming we have confluent and terminating rewrite rules. However, this is clearly not the case:

```agda
{-# OPTIONS --rewriting --confluence-check #-}

open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Agda.Builtin.Unit

postulate
  f : {X : Set} → X → Set
  g : {X : Set} → Set → X

eq : (A : Set) → f (g A) ≡ f tt
eq A = refl

postulate
  rew-fg : {X : Set} (A : Set) → f (g {X} A) ≡ A
{-# REWRITE rew-fg #-}

cast : {A B : Set} → A → B
cast {A} {B} x with eq A | eq B
... | refl | refl = x
```
Here, `cast` is definitionally an identity function between any two types, e.g. `cast 42 : Bool` evaluates to `42 : Bool`.

**Comment by jespercockx** [comment]
Well, anyway, I do not know how to solve this problem, so for now we are back to the situation we had before, which is that you should use rewrite rules on your own risk, and if you do it wrong then anything can and will happen except for internal errors. I mainly created this issue to document the behaviour and in case someone ever finds a more satisfying solution.

**Comment by NathanielB123** [comment]
My instinct for how to fix this would be to make `nlPatVarsUnder` type-directed, and not consider variables to be bound on the LHS if the type of the expression they occur in is neutral (i.e. neutral types after a substitution could reduce to unit). In other words, we should not consider variables bound on the LHS if the surrounding expression is possibly a definitional-singleton. Not sure how many rewrite rules this would break.

My justification is based on how another source of definitional singularity, specifically `Prop`, is handled correctly:
```agda
postulate
  f : {X : Prop} → X → Set
  g : {X : Prop} → Set → X

postulate
  rew-fg : {X : Prop} (A : Set) → f {X} (g {X} A) ≡ A
{-# REWRITE rew-fg #-}

-- warning: -W[no]RewriteVariablesNotBoundByLHS
-- rew-fg  is not a legal rewrite rule, since the following variables are not bound by the left hand side:  A
-- when checking the pragma REWRITE rew-fg
```

In practice, this comes from how `Prop`-sorted expressions are turned into `PTerm`s which cannot bind variables in rewrite rule matching. Perhaps also turning expressions of neutral type into `PTerm`s could work?
